### PR TITLE
Smaller tables

### PR DESCRIPTION
**Pull request by benibela**
See #67

### CODE REVIEW

**Comment by ivarne** [comment]
Seems like travis is failing because of an expired cert
```
# use JuliaLang caching (https://github.com/staticfloat/cache.julialang.org)
# so that Travis builds do not depend on anyone's flaky servers but our own
    - URLCACHE=https://cache.e.ip.saba.us/
```

cc: @staticfloat

**Comment by ivarne** [comment]
The travis failiure will probably be fixed by https://github.com/JuliaLang/utf8proc/pull/69

**Comment by tkelman** [comment]
It might need a rebase and force push though. Changes to `.travis.yml` might not propagate automatically by just restarting the build.

**Comment by stevengj** [comment]
Needs a version bump.  Since this is a technically a backwards incompatible change, we'll need to change the major version? 

Maybe we should take the opportunity to export other accessor functions, like `utf8proc_casefold(codepoint)`, `utf8proc_bidi(codepoint)`, ...?

We could unexport the `utf8proc_property_struct` entirely, replacing it by accessor functions.  Theoretically, this would make it less efficient to extract multiple properties of a codepoint at once, but I'm skeptical that this is common in practice.

**Comment by stevengj** [comment]
is this ready to merge?

**Comment by benibela** [comment]
Unless you want to increase the version number first for people using the property struct directly.

And Unicode 9 was just released. Might be worth checking if it still fits in the tables. Some are almost full.

**Comment by stevengj** [comment]
Needs a rebase and update now that Unicode 9 support has been merged.

File: utf8proc.h [hunk 1]
```diff
@@ -550,6 +549,13 @@ UTF8PROC_DLLEXPORT utf8proc_int32_t utf8proc_tolower(utf8proc_int32_t c);
 UTF8PROC_DLLEXPORT utf8proc_int32_t utf8proc_toupper(utf8proc_int32_t c);
 
 /**
+ * Given a codepoint `c`, return the codepoint of the corresponding
+ * title-case character, if any; otherwise (if there is no title-case
+ * variant, or if `c` is not a valid codepoint) return `c`.
+ */
+UTF8PROC_DLLEXPORT utf8proc_int32_t utf8proc_totitle(utf8proc_int32_t c);
```
Comment (stevengj) [inline utf8proc.h]: Should this be equivalent to `toupper` if there is no titlecase?

**Comment by stevengj** [comment]
Awesome, thanks for updating this.  Did you have any problems with the table size?

**Comment by stevengj** [comment]
What's the overall reduction in (compiled) table size from this patch?

**Comment by benibela** [comment]
200 k (42%)

File: data/data_generator.rb [hunk 2]
```diff
@@ -115,22 +115,52 @@ def str2c(string, prefix)
...
+def cpary2c(array)
+  return "UINT16_MAX" if array.nil? || array.length == 0
+  lencode = array.length - 1 #no sequence has len 0, so we encode len 1 as 0, len 2 as 1, ... 
+  array = cpary2utf16encoded(array)
+  if lencode >= 7 #we have only 3 bits for the length (which is already cutting it close. might need to change it to 2 bits in future Unicode versions)
+    array = [lencode] + array 
+    lencode = 7
+  end  
+  idx = pushary(array) 
+  raise "Array index out of bound" if idx > 0x1FFF
```
Comment (stevengj) [inline data/data_generator.rb:147]: This exception triggered in Unicode 14.   @benibela, what do you suggest?

Comment (benibela) [reply data/data_generator.rb:147]: @stevengj one bit needs to be taken from lencode (7 -> 3) and given to idx (0x1FFF -> 0x3FFF, 13 -> 14)

and the same in seqindex_write_char_decomposed

Comment (stevengj) [reply data/data_generator.rb:147]: Could you make a PR?

**Pull request merged**